### ISSUE TITLE
XMLLoggerTest.java to use verifyWithInlineConfigParserAndXmlLogger

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#16360]
After merge of https://github.com/checkstyle/checkstyle/pull/16314

We have ability to update all/most test methods in [XMLLoggerTest.java](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/XMLLoggerTest.java) to use `verifyWithInlineConfigParserAndXmlLogger`

Goal: use it as much as possible, ideally in all test methods.

Good example of expected update https://github.com/checkstyle/checkstyle/pull/16701/files

**Comment by roulpriya** [comment]
@romani, I am working on this issue. 
I see three potential test methods in `XMLLoggerTest.java` that could use `verifyWithInlineConfigParserAndXmlLogger`
- `testAddError`: Adding any error to the source
- `testAddIgnored`: Adding a test which is ignored
- `testAddErrorOnZeroColumns`: Adding a test that fails on the entire line like `EmptyLineSeparatorCheck`

I have raised a draft PR fixing `testAddError`. I wanted to validate if I understand this correctly.

**Comment by romani** [comment]
Your draft PR is in right direction.
Please do all test methods migration in separate PRs.

**Comment by roulpriya** [comment]
I am not able to get this error: https://app.circleci.com/pipelines/github/checkstyle/checkstyle/32027/workflows/2d0479fd-798e-4a53-b962-2c0ed35f6bb3/jobs/862027/parallel-runs/0/steps/0-103?invite=true#step-103-263616_54

This says `Could not find artifact org.xwiki.commons:xwiki-commons`

Can you help me with this?

**Comment by roulpriya** [comment]
In the `XMLLoggerTest.testAddIgnored()` case, which checkstyle check should I use? Do I have to add suppression to ignore the checks?

**Comment by romani** [comment]
You need to use some Check in xml config that use severity ignore 
https://github.com/checkstyle/checkstyle/blob/359fa919d6b96ea334346459b77b7e54931dee65/src/test/java/com/puppycrawl/tools/checkstyle/XMLLoggerTest.java#L256

**Comment by gianmarcoschifone** [comment]
@romani I'm not sure on how to update the `testAddErrorWithNullFileName()` since it would require an `AuditEvent` with null `fileName`. Could you please guide me?

**Comment by romani** [comment]
>since it would require an AuditEvent with null fileName. Could you please guide me?

Yes, not possible if all other code works well, please add comment above test method to explain why we can not use input based approach

**Comment by vivek-0509** [comment]
Hey @romani,

I went through the remaining test methods and found that several of them cannot be safely migrated to use `verifyWithInlineConfigParserAndXmlLogger` due to the nature of what they validate.

In particular, they rely on:

- Manually constructed `AuditEvent` instances (including `null` file names).
- Custom `OutputStream` implementations to assert `close()` behavior.
- Direct control over the ordering of logger method calls.
- Validation of constructor behavior and internal fields (e.g., `closeStream`).

Because of these constraints, these scenarios cannot be reliably reproduced through the inline config verifier without redesigning the helper.

The following methods fall into this category and likely need to remain as direct unit tests, with short explanatory comments added to clarify why they were not migrated:

- `testAddException()`
- `testAddExceptionAfterFileStarted()`
- `testAddExceptionBeforeFileFinished()`
- `testAddExceptionBetweenFileStartedAndFinished()`
- `testAuditFinishedWithoutFileFinished()`
- `testNullOutputStreamOptions()`
- `testFinishLocalSetup()`
- `testCtorWithTwoParametersCloseStreamOptions()`
- `testCtorWithTwoParametersNoneStreamOptions()`

Could you please confirm if I’m thinking in the right direction, and if adding explanatory comments to these tests (and leaving them as-is) would be the right approach?

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/62c038ddf9de15f2d9f4b5b9c2ca97ddac72c197/src/test/java/com/puppycrawl/tools/checkstyle/XMLLoggerTest.java#L258

You can try to run checkstyle on non compiled java Input  and there will exception .


>testNullOutputStreamOptions()
testFinishLocalSetup()
testCtorWithTwoParametersCloseStreamOptions()
testCtorWithTwoParametersNoneStreamOptions()

This likely can be only pure unit testing

**Comment by vivek-0509** [comment]
Thanks for the clarification @romani.

Understood , I will keep the constructor and lifecycle tests (`testNullOutputStreamOptions`, `testFinishLocalSetup`, and the ctor-related tests) as pure unit tests and add short comments explaining why they cannot be migrated.

For the `addException*` tests, I will try to reproduce the behavior by running Checkstyle on intentionally invalid / non compilable Java inputs so that real exceptions are triggered and logged through the inline verifier.

**Comment by Youssefalaa7** [comment]
Hey @romani 

While exploring the XMLLogger tests, I noticed something in XMLLoggerTest.testAuditFinishedWithoutFileFinished().

Issue: The test name says "WithoutFileFinished" but the code actually calls fileFinished() on line 364. Without this fix, it only tests adding an error (the normal happy path), not the edge case of missing fileFinished().

Why it matters: XMLLogger buffers errors in memory when fileStarted() is called, and only writes them to XML when fileFinished() is called. If fileFinished() is skipped, the buffered errors are never written.

Fix:

Remove line 364: logger.fileFinished(errorEvent);
Change expected XML from ExpectedXMLLoggerError.xml to ExpectedXMLLoggerEmpty.xml (since no errors are written without fileFinished)
Add JavaDoc comment explaining why this test can't use verifyWithInlineConfigParserAndXmlLogger() (since we need to intentionally skip fileFinished())
Without this fix, the test doesn't verify the edge case it claims to test - it only tests adding an error. Also, this test cannot be converted to use verifyWithInlineConfigParserAndXmlLogger() because that helper would automatically call fileFinished() as part of the normal audit lifecycle.

If this looks correct, I'll submit a PR.

**Comment by romani** [comment]
It might have outdated name.
Please convert to verifyWithInlineConfigParserAndXmlLogger, let's see what code becomes not covered, send PR to see all details

**Comment by Adam-Gamil** [comment]
I am on testFileOpenTag functions

**Comment by yn-mi12** [comment]
Hello, is this issue still open?

**Comment by romani** [comment]
Yes https://github.com/checkstyle/checkstyle/blob/b19b3e3b987c71304e9f09dec4bec0194a75c11c/src/test/java/com/puppycrawl/tools/checkstyle/XMLLoggerTest.java#L278
And down, there are still 4 methods to be updated

**Comment by yn-mi12** [comment]
Great, I will work on it.

**Comment by yn-mi12** [comment]
I started working on the issue and refactored the tests testAddExceptionAfterFileStarted, testAddExceptionBeforeFileFinished, and testAddExceptionBetweenFileStartedAndFinished. However, currently, the expected output is an error instead of an exception, which violates the test coverage requirements for XMLLogger. For example:
``` java 
/*xml
<module name="Checker">
    <module name="TreeWalker">
        <module name="MemberName"/>
    </module>
</module>
*/
package com.puppycrawl.tools.checkstyle.xmllogger;

public class InputXMLLoggerExceptionAfterFileStarted {
    private int BadName;
    private int goodName;
}
```
``` xml
<?xml version="1.0" encoding="UTF-8"?>
<checkstyle version="">
<file name="InputXMLLoggerExceptionAfterFileStarted.java">
<error line="11" column="15" severity="error" message="Name 'BadName' must match pattern '^[a-z][a-zA-Z0-9]*$'." source="com.puppycrawl.tools.checkstyle.checks.naming.MemberNameCheck"/>
</file>
</checkstyle>
```
I am not sure how to trigger a Runtime Exception without explicitly calling `logger.addException()`. Should I write extra tests specifically for writeFileMessages() and addException() in XMLLogger, or is there another solution?

**Comment by romani** [comment]
@yn-mi12,
Please look at main code where we use `addException` and try to see what bad we can do to make it executed.
quick search shows https://github.com/checkstyle/checkstyle/blob/096065b613d4de39c0df4d4cafd086bf276cf317/src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java#L230-L237

so exception should be any and it there should not be anything else as violations in such file.

you can cause excpetion from Checkstyle by placing un-compiled Input java file so to cause our parser to fail with exception.
We already have a lot of examples on this, look around

**Comment by yn-mi12** [comment]
@romani 
Thank you for your reply. I looked at previous commits and similar tests, however none of those I saw use `<exception><exception/>`, only `<error><error/>`. I figured the problem may be that `verifyWithInlineConfigParserAndXmlLogger()` uses `Checker.java`, which doesn't call the `addException()` method anywhere, only `addError()`.  In `TreeWalker.java`, the exception is thrown or converted to Violation:
``` java
catch (Exception exc) {
                if (!skipFileOnJavaParseException) {
                    throw exc;
                }
                skip = true;
                violations.add(new Violation(1, Definitions.CHECKSTYLE_BUNDLE, PARSE_EXCEPTION_MSG,
                            new Object[] {exc.getMessage()}, javaParseExceptionSeverity, null,
                            getClass(), null));
                addViolations(violations);
            }
```
Then in `Checker.java`, violations trigger `addError()`:
``` java
for (final Violation element : errors) {
            final AuditEvent event = new AuditEvent(this, stripped, element);
            if (filters.accept(event)) {
                hasNonFilteredViolations = true;
                for (final AuditListener listener : listeners) {
                    listener.addError(event);
                }
            }
        }
```

That's why I believe it's not possible to create tests that produce XMLLogger exception output without directly calling `logger.addException()` in `XMLLoggerTest.java`, or adding an `addException()` listener in `Checker.java`.

 It is very possible I am missing something. Could you please explain in more detail what you meant?

**Comment by romani** [comment]
>verifyWithInlineConfigParserAndXmlLogger() uses Checker.java, which doesn't call the addException() method anywhere, only addError(). In TreeWalker.java, the exception is thrown or converted to Violation

thanks a lot, so lets add comment above method to explain why verifyxxxxx can not be used.
All methods should use verifyxxxxxxx or has comment that explains specific situation.
We potentially can make test Rootmodule that uses this function of XMLLogger, example https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/testmodules/TestRootModuleChecker.java , but current main Root module does not do this.

**Comment by romani** [comment]
Final fix is merged
